<!DOCTYPE html>
<html>

<head>

<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>page5</title>


<style type="text/css">
body {
  font-family: Helvetica, arial, sans-serif;
  font-size: 14px;
  line-height: 1.6;
  /*padding-top: 10px;
  padding-bottom: 10px; */
  background-color: white;
  /* padding: 30px;  */ }

body > *:first-child {
  margin-top: 0 !important; }
body > *:last-child {
  margin-bottom: 0 !important; }

a {
  color: #4183C4; }
a.absent {
  color: #cc0000; }
a.anchor {
  display: block;
  padding-left: 30px;
  margin-left: -30px;
  cursor: pointer;
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0; }

h1, h3, h3, h4, h5, h6 {
  margin: 20px 0 10px;
  padding: 0;
  font-weight: bold;
  -webkit-font-smoothing: antialiased;
  cursor: text;
  position: relative; }

h1:hover a.anchor, h3:hover a.anchor, h3:hover a.anchor, h4:hover a.anchor, h5:hover a.anchor, h6:hover a.anchor {
  background: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA09pVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMy1jMDExIDY2LjE0NTY2MSwgMjAxMi8wMi8wNi0xNDo1NjoyNyAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvIiB4bWxuczp4bXBNTT0iaHR0cDovL25zLmFkb2JlLmNvbS94YXAvMS4wL21tLyIgeG1sbnM6c3RSZWY9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9zVHlwZS9SZXNvdXJjZVJlZiMiIHhtcDpDcmVhdG9yVG9vbD0iQWRvYmUgUGhvdG9zaG9wIENTNiAoMTMuMCAyMDEyMDMwNS5tLjQxNSAyMDEyLzAzLzA1OjIxOjAwOjAwKSAgKE1hY2ludG9zaCkiIHhtcE1NOkluc3RhbmNlSUQ9InhtcC5paWQ6OUM2NjlDQjI4ODBGMTFFMTg1ODlEODNERDJBRjUwQTQiIHhtcE1NOkRvY3VtZW50SUQ9InhtcC5kaWQ6OUM2NjlDQjM4ODBGMTFFMTg1ODlEODNERDJBRjUwQTQiPiA8eG1wTU06RGVyaXZlZEZyb20gc3RSZWY6aW5zdGFuY2VJRD0ieG1wLmlpZDo5QzY2OUNCMDg4MEYxMUUxODU4OUQ4M0REMkFGNTBBNCIgc3RSZWY6ZG9jdW1lbnRJRD0ieG1wLmRpZDo5QzY2OUNCMTg4MEYxMUUxODU4OUQ4M0REMkFGNTBBNCIvPiA8L3JkZjpEZXNjcmlwdGlvbj4gPC9yZGY6UkRGPiA8L3g6eG1wbWV0YT4gPD94cGFja2V0IGVuZD0iciI/PsQhXeAAAABfSURBVHjaYvz//z8DJYCRUgMYQAbAMBQIAvEqkBQWXI6sHqwHiwG70TTBxGaiWwjCTGgOUgJiF1J8wMRAIUA34B4Q76HUBelAfJYSA0CuMIEaRP8wGIkGMA54bgQIMACAmkXJi0hKJQAAAABJRU5ErkJggg==) no-repeat 10px center;
  text-decoration: none; }

h1 tt, h1 code {
  font-size: inherit; }

h3 tt, h3 code {
  font-size: inherit; }

h3 tt, h3 code {
  font-size: inherit; }

h4 tt, h4 code {
  font-size: inherit; }

h5 tt, h5 code {
  font-size: inherit; }

h6 tt, h6 code {
  font-size: inherit; }

h1 {
  font-size: 28px;
  color: black; }

h3 {
  font-size: 24px;
  border-bottom: 1px solid #cccccc;
  color: black; }

h3 {
  font-size: 18px; }

h4 {
  font-size: 16px; }

h5 {
  font-size: 14px; }

h6 {
  color: #777777;
  font-size: 14px; }

p, blockquote, ul, ol, dl, li, table, pre {
  margin: 15px 0; }

hr {
  background: transparent url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAYAAAAECAYAAACtBE5DAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAAyJpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvIiB4bWxuczp4bXBNTT0iaHR0cDovL25zLmFkb2JlLmNvbS94YXAvMS4wL21tLyIgeG1sbnM6c3RSZWY9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9zVHlwZS9SZXNvdXJjZVJlZiMiIHhtcDpDcmVhdG9yVG9vbD0iQWRvYmUgUGhvdG9zaG9wIENTNSBNYWNpbnRvc2giIHhtcE1NOkluc3RhbmNlSUQ9InhtcC5paWQ6OENDRjNBN0E2NTZBMTFFMEI3QjRBODM4NzJDMjlGNDgiIHhtcE1NOkRvY3VtZW50SUQ9InhtcC5kaWQ6OENDRjNBN0I2NTZBMTFFMEI3QjRBODM4NzJDMjlGNDgiPiA8eG1wTU06RGVyaXZlZEZyb20gc3RSZWY6aW5zdGFuY2VJRD0ieG1wLmlpZDo4Q0NGM0E3ODY1NkExMUUwQjdCNEE4Mzg3MkMyOUY0OCIgc3RSZWY6ZG9jdW1lbnRJRD0ieG1wLmRpZDo4Q0NGM0E3OTY1NkExMUUwQjdCNEE4Mzg3MkMyOUY0OCIvPiA8L3JkZjpEZXNjcmlwdGlvbj4gPC9yZGY6UkRGPiA8L3g6eG1wbWV0YT4gPD94cGFja2V0IGVuZD0iciI/PqqezsUAAAAfSURBVHjaYmRABcYwBiM2QSA4y4hNEKYDQxAEAAIMAHNGAzhkPOlYAAAAAElFTkSuQmCC) repeat-x 0 0;
  border: 0 none;
  color: #cccccc;
  height: 4px;
  padding: 0;
}

body > h3:first-child {
  margin-top: 0;
  padding-top: 0; }
body > h1:first-child {
  margin-top: 0;
  padding-top: 0; }
  body > h1:first-child + h3 {
    margin-top: 0;
    padding-top: 0; }
body > h3:first-child, body > h4:first-child, body > h5:first-child, body > h6:first-child {
  margin-top: 0;
  padding-top: 0; }

a:first-child h1, a:first-child h3, a:first-child h3, a:first-child h4, a:first-child h5, a:first-child h6 {
  margin-top: 0;
  padding-top: 0; }

h1 p, h3 p, h3 p, h4 p, h5 p, h6 p {
  margin-top: 0; }

li p.first {
  display: inline-block; }
li {
  margin: 0; }
ul, ol {
  padding-left: 30px; }

ul :first-child, ol :first-child {
  margin-top: 0; }

dl {
  padding: 0; }
  dl dt {
    font-size: 14px;
    font-weight: bold;
    font-style: italic;
    padding: 0;
    margin: 15px 0 5px; }
    dl dt:first-child {
      padding: 0; }
    dl dt > :first-child {
      margin-top: 0; }
    dl dt > :last-child {
      margin-bottom: 0; }
  dl dd {
    margin: 0 0 15px;
    padding: 0 15px; }
    dl dd > :first-child {
      margin-top: 0; }
    dl dd > :last-child {
      margin-bottom: 0; }

blockquote {
  border-left: 4px solid #dddddd;
  padding: 0 15px;
  color: #777777; }
  blockquote > :first-child {
    margin-top: 0; }
  blockquote > :last-child {
    margin-bottom: 0; }

table {
  padding: 0;border-collapse: collapse; }
  table tr {
    border-top: 1px solid #cccccc;
    background-color: white;
    margin: 0;
    padding: 0; }
    table tr:nth-child(2n) {
      background-color: #f8f8f8; }
    table tr th {
      font-weight: bold;
      border: 1px solid #cccccc;
      margin: 0;
      padding: 6px 13px; }
    table tr td {
      border: 1px solid #cccccc;
      margin: 0;
      padding: 6px 13px; }
    table tr th :first-child, table tr td :first-child {
      margin-top: 0; }
    table tr th :last-child, table tr td :last-child {
      margin-bottom: 0; }

/* img {
  max-width: 100%; } */

span.frame {
  display: block;
  overflow: hidden; }
  span.frame > span {
    border: 1px solid #dddddd;
    display: block;
    float: left;
    overflow: hidden;
    margin: 13px 0 0;
    padding: 7px;
    width: auto; }
  span.frame span img {
    display: block;
    float: left; }
  span.frame span span {
    clear: both;
    color: #333333;
    display: block;
    padding: 5px 0 0; }
span.align-center {
  display: block;
  overflow: hidden;
  clear: both; }
  span.align-center > span {
    display: block;
    overflow: hidden;
    margin: 13px auto 0;
    text-align: center; }
  span.align-center span img {
    margin: 0 auto;
    text-align: center; }
span.align-right {
  display: block;
  overflow: hidden;
  clear: both; }
  span.align-right > span {
    display: block;
    overflow: hidden;
    margin: 13px 0 0;
    text-align: right; }
  span.align-right span img {
    margin: 0;
    text-align: right; }
span.float-left {
  display: block;
  margin-right: 13px;
  overflow: hidden;
  float: left; }
  span.float-left span {
    margin: 13px 0 0; }
span.float-right {
  display: block;
  margin-left: 13px;
  overflow: hidden;
  float: right; }
  span.float-right > span {
    display: block;
    overflow: hidden;
    margin: 13px auto 0;
    text-align: right; }

code, tt {
  margin: 0 2px;
  padding: 0 5px;
  white-space: nowrap;
  border: 1px solid #eaeaea;
  background-color: #f8f8f8;
  border-radius: 3px; }

pre code {
  margin: 0;
  padding: 0;
  white-space: pre;
  border: none;
  background: transparent; }

.highlight pre {
  background-color: #f8f8f8;
  border: 1px solid #cccccc;
  font-size: 13px;
  line-height: 19px;
  overflow: auto;
  padding: 6px 10px;
  border-radius: 3px; }

pre {
  background-color: #f8f8f8;
  border: 1px solid #cccccc;
  font-size: 13px;
  line-height: 19px;
  overflow: auto;
  padding: 6px 10px;
  border-radius: 3px; }
  pre code, pre tt {
    background-color: transparent;
    border: none; }

sup {
    font-size: 0.83em;
    vertical-align: super;
    line-height: 0;
}

kbd {
  display: inline-block;
  padding: 3px 5px;
  font-size: 11px;
  line-height: 10px;
  color: #555;
  vertical-align: middle;
  background-color: #fcfcfc;
  border: solid 1px #ccc;
  border-bottom-color: #bbb;
  border-radius: 3px;
  box-shadow: inset 0 -1px 0 #bbb
}

* {
	-webkit-print-color-adjust: exact;
}
/* @media screen and (min-width: 914px) {
    body {
        width: 854px;
        margin:0 auto;
    }
} */
@media print {
	table, pre {
		page-break-inside: avoid;
	}
	pre {
		word-wrap: break-word;
	}
  body {
    padding: 2cm; 
  }
}
</style>


</head>

<body>
  <h2>5.1 Confidential Bioinformatics Pipeline Architecture</h2>

  <h3 id="overview">5.1.1 Overview</h3>
  <p>
    This chapter presents a hands-on implementation of <strong>Confidential Computing</strong> applied to bioinformatics pipelines. By combining <strong>Dockerized workflow execution</strong> with the <strong>Certifier Framework</strong>, the system ensures that sensitive genomic data is processed only within <em>trusted execution environments (TEEs)</em>. Users interact with this setup via a Python-based orchestration script, <code>control_docker.py</code>, which automates the instantiation and execution of both <strong>Server VMs</strong> (handling security policies and attestation) and <strong>Client VMs</strong> (executing domain-specific workflows like FastQC, BWA/SAMtools, and GATK).
  </p>
  <p>
    This architecture not only enforces data confidentiality and computational integrity but also demonstrates a practical, scalable framework for <em>secure genomic analysis across cloud or edge environments</em>. The workflows are modular, configurable, and allow researchers to maintain compliance with strict privacy standards while automating the entire pipeline lifecycle.
  </p>
  <h3 id="pipeline-architecture">5.1.2 Pipeline Architecture</h3>
  <ul>
    <li><strong>Server VM:</strong> Runs the certifier service that manages attestation and policy enforcement</li>
    <li><strong>Client VM:</strong> Performs one of four bioinformatics workflows on genomic data</li>
  </ul>


  <div style="text-align: center;">
    <p align = "center"><img src="/MizzouCloudDevOps/img/lab9/step6/pipeline.png" width=700></p>
  
    <p style="max-width:800px; margin: 20px auto 0; font-size: 14px; color: #333;">
      <strong>Figure 5.1:</strong> This architecture illustrates a secure bioinformatics pipeline using the Certifier Framework. The User initiates operations through a Python Controller that orchestrates Server and Client VMs. The Server VM handles security policies via the Certifier Service, while the Client VM runs selected workflows (FastQC, BWA/SAMtools, or GATK) on genomic data. Outputs like HTML Reports, Sorted BAM, and VCF files are generated while maintaining data confidentiality through secure gRPC communication and Trusted Execution Environments.
    </p>
  </div>

  <h2>5.2 Instance Preparation</h2>
  <p>
  Let us start creating our instance for server VM and client VM, similar to that of chapter 2
  </p>

  <h3>5.2.1 In EC2 Dashboard, click on <strong>Instances</strong> in the left menu.</h3>
  <p align="center"><img src="/MizzouCloudDevOps/img/lab9/step6/select_instances.png" width=250></p>

  <h3>5.2.2 Server Instance</h2>
  <p>Since we are already working with an AWS instance in out earlier chapters. We will repurpose the same instance as a server for our use case. Therefore, let us rename the instance in a way that will be easier for us later on. 
  Select your instance and enter a unique name. (e.g., <code>secure-server-MCD</code>)</p>
  <p align="center"><img src="/MizzouCloudDevOps/img/lab9/step6/rename_server.png" width=700></p>
  <p>Since, the server is setup now. We need to create another instance which work as a client.</p>

  <h3>5.2.3 In EC2 Dashboard, click on <strong>AMIs</strong> in the left menu.</h3>
  <p align="center"><img src="/MizzouCloudDevOps/img/lab9/step6/AMI_option.png" width=250></p>

  <h3>5.2.4 Search: <code>certifier_fw_image</code>, filter by <strong>Public Images</strong>. Select and click <strong>Launch instance from AMI</strong></h3>
  <p align="center"><img src="/MizzouCloudDevOps/img/lab9/step6/choose_AMI.png" width=700></p>

  <h3>5.2.5 Create a Client Instance</h2>
  <p>We will now create another instance which be work as a client. Enter a unique name (e.g., <code>secure-client-MCD</code>)  under "Name and tags". </p>
  <p align="center"><img src="/MizzouCloudDevOps/img/lab9/step6/secure-client.png" width=700></p>
  <p>We will select the Instance Type as <code>t2.medium</code> similar to how we did in Chapter 2.</p>
  <p align="center"><img src="/MizzouCloudDevOps/img/lab9/step6/select_t2_medium.png" width=700></p>

  <p>You can use the same key pair from the previous chapter.</p>
  <p align="center"><img src="/MizzouCloudDevOps/img/lab9/step6/select_key.png" width=700></p>
  <p>Or create a new one.</p>
  <p align="center"><img src="/MizzouCloudDevOps/img/lab9/step6/name_key_pair.png" width=700></p>
  
  <blockquote style = "background-color: #e5f5e0 ; margin: 0 ; font-family: calibri;border: 2px solid #00441b; border-left: 6px solid #00441b; padding : 0.5em; border-radius: 8px; background-style:discrete" >
    <font size = 3 color = green> <b> Note: </b></font><font color = 'black'>
    The created key pair will be downloaded locally.
    </font> </blockquote> 
  <p>Review instance summary on the left and click <strong>Launch Instance</strong></p>
  
  
  <!--<h3>5.2.4 Create a Client Instance</h2>
  <p>Follow all the steps from 5.2.3 only exception will be the instance name (e.g., <code>secure-client-MCD</code>) under "Name and tags"</p>
  <p align="center"><img src="/MizzouCloudDevOps/img/lab9/step6/secure-client.png" width=700></p>

  <p>Review instance summary on the left and click <strong>Launch Instance</strong>.</p>
  <p align="center"><img src="lab9/vera.png" width=700></p>-->

  <h3>5.2.6 Launching Instances</h2>
  <p>If you go back to the EC2 dashboard, you can see two instances that we recently created.</p>
  <p align="center"><img src="/MizzouCloudDevOps/img/lab9/step6/instances.png" width=700></p>

  <p>Click on your client instance. You will see the client instance summary. Keep note of <strong>Public IPv4 Address</strong> and <strong>Private IPv4 address</strong>. Now click on the <strong>Connect</strong> button in the top right corner.</p>
  <p align="center"><img src="/MizzouCloudDevOps/img/lab9/step6/instance_summary.png" width=700></p>

  <p>You will see a Connect to Instance page, click on <strong>Connect</strong>.</p>
  <p align="center"><img src="/MizzouCloudDevOps/img/lab9/step6/final_con.png" width=700></p>

  <p><strong>Your AWS instance:</strong> If you face a NAT/connection issue, reload or relaunch. You should see the following view:</p>
  <p align="center"><img src="/MizzouCloudDevOps/img/lab9/step6/instance_page.png" width=700></p>
  <p>If you failed to note the <strong>Public IPv4 address</strong> and <strong>Private IPv4 address</strong> from the previous steps. You can also find those on the bottom of the instance window</p>

  <p>Repeat the steps 5.2.6 for the server instance, note the <strong>Public IPv4 address</strong> and <strong>Private IPv4 address</strong>, and click on Connect. You will now see two instances in your browser tabs.</p>

  <h3>5.2.7 Configuring Security Group</h2>
  <p>We want to allow the client VM to connect with the server VM. For that, we need to allow the client VM's IP Address as an inbound rule on the Server's security group.</p>
  <p>Go back to the EC2 Dashboard. Click on the server instance.<p>
  <p align="center"><img src="/MizzouCloudDevOps/img/lab9/step6/server_sg1.png" width=700></p>

  <p>Scroll down, click on the <strong>Security</strong> tab. Click on the <strong>Security Groups</strong></p>
  <p align="center"><img src="/MizzouCloudDevOps/img/lab9/step6/security_sg2.png" width=700></p>


  <p>You will see a security configuration wizard on a new tab. Click on <strong>Edit Inbound Rules</strong>.</p>
  <p align="center"><img src="/MizzouCloudDevOps/img/lab9/step6/security_inbound1.png" width=700></p>

  <p>You will see a security configuration wizard on a new tab. Click on <strong>Edit Inbound Rules</strong>.</p>
  <ol>
  <li>Step 1: Click <strong>Add rule</strong> to insert a new, empty inbound‑rule row into the security group.</li>
  <li>Step 2: In the first new row, set <strong>Type</strong> to <strong>Custom TCP</strong>, enter <strong>8123</strong> as the <strong>Port range</strong>, and choose the  <strong>Client VM’s Private IP/32</strong> that we noted in previous steps, (e.g. 172.31.31.181/32) as the <strong>Source</strong>.</li>
  <li>Step 3: Click <strong>Add rule</strong> again. In the second new row, again select <strong>Custom TCP</strong>, enter <strong>8124</strong> as the <strong>Port range</strong>, and set the same <strong>Client VM’s Private IP/32</strong> , (e.g. 172.31.31.181/32) as the <strong>Source</strong>.</li>
  <li>Step 4: Click <strong>Save rules</strong> to apply these changes and allow traffic from the specified IP on ports 8123 and 8124.</li>
  <p align="center"><img src="/MizzouCloudDevOps/img/lab9/step6/adding_rules.png" width=700></p>
  <p>Now we are done setting up our instances.</p>



  <h2 id="bioinformatics-workflows">5.3 Bioinformatics Workflows</h2>

  <h3 id="1-sequence-quality-control-fastqc">5.3.1 Sequence Quality Control (FastQC)</h3>
<p>
  The purpose of this step is to assess the overall quality of raw sequencing data in order to identify potential issues such as low-quality base calls, adapter contamination, and sequence duplication levels. In this case, the input file used is SRR2584863_1.fastq, which is a raw FASTQ file containing sequencing reads. The specific command used for this step is yet to be specified.
</p>

<div><pre><button class="copyButton" ng-click="copyToClipboard('2db354a0-c14a-40dc-87d6-d8e6863f7b37')" type="button"><i class="fa fa-files-o"></i></button>
<code class="language-bash">
$ fastqc SRR2584863_1.fastq --threads 2 --quiet -o /results</code>
<code hidden id="2db354a0-c14a-40dc-87d6-d8e6863f7b37">
fastqc SRR2584863_1.fastq --threads 2 --quiet -o /results
</code></pre></div>
  
  <h3>5.3.2 Read Alignment (BWA + SAMtools)</h3>
  <p>In this step, sequencing reads are aligned to a reference genome to determine their genomic origin. This is a crucial foundational step in genomics, as it enables researchers to accurately map reads and identify genomic variants. Such alignment is essential for downstream processes including variant calling, gene expression quantification, and the detection of structural variations. The required inputs include a reference genome in FASTA format and paired-end reads in FASTQ format (e.g., SRR2584863_1.fastq, SRR2584863_2.fastq). The output is a sorted and indexed BAM file that contains the aligned reads, which are generated through the BWA-MEM algorithm and then processed using SAMtools. BWA offers fast and memory-efficient alignment, while SAMtools supports efficient sorting, indexing, and conversion of the alignment data for further downstream analysis.</p>
  
  <h3>5.3.3 Indel Calling (GATK)</h3>
  <p>This step focuses on detecting small insertions and deletions (indels) in the sample genome relative to the reference genome. Indels can have significant functional consequences, such as introducing frameshifts or disrupting genes, making their detection critical for studies on rare diseases, cancer genomics, and population genetics. The required inputs are a BAM file of aligned reads and the reference genome, and the output is a VCF file containing the identified and filtered indel variants. The GATK pipeline conducts local realignment and probabilistic variant detection using tools like HaplotypeCaller or Mutect2. Post-calling filters are applied to retain only high-confidence indels based on features such as depth, base quality, and read context.</p>
  
  <h3>5.3.4 SNP Calling (GATK)</h3>
  <p>This step identifies single nucleotide polymorphisms (SNPs), which are the most common form of genetic variation. SNPs are vital for genome-wide association studies (GWAS), personalized medicine, and evolutionary research, as their genomic positions can affect gene regulation, protein function, or serve as biomarkers. The input for this step includes a BAM file with aligned reads and the corresponding reference genome. The output is a filtered VCF file that includes SNPs along with associated annotations such as genotype, quality scores, and depth. The GATK toolkit performs SNP discovery using statistical models to distinguish true variants from sequencing errors. It also applies filtering criteria—such as mapping quality, strand bias, and allele balance—to improve the precision of the results.</p>
  

  <h2>5.4 Controlling the Pipeline</h2>
  <p>All control is done through <code>control_docker.py</code>. Here's how to use it:</p>

<div>
  <pre><button ng-click="copyToClipboard('cmd-2')" type="button" class="copyButton"><i class="fa fa-files-o"></i></button>
  <code class="language-bash">$ python control_docker.py [GLOBAL_ARGS] COMMAND [COMMAND_ARGS]</code>
  <code id="cmd-2" hidden>python control_docker.py [GLOBAL_ARGS] COMMAND [COMMAND_ARGS]</code>
  </pre>
</div>

<p>We will be able to monitor the process</p>
<ul>
  <li>Real-time updates in terminal</li>
  <li>Server logs: attestation and policy enforcement</li>
  <li>Client logs: workflow execution</li>
</ul>

  <h3>5.4.1 Activate Virtual Environment</h3>
  <p>We will begin from our server instance. Navigate to the server VM on your browser tab.</p>
  <p>Before running any commands, the virtual environment needs to be activated to ensure that all dependencies are correctly sourced:</p>

  <div>
  <pre><button ng-click="copyToClipboard('cmd-3')" type="button" class="copyButton"><i class="fa fa-files-o"></i></button>
  <code class="language-bash">$ cd /home/ubuntu</code>
  <code id="cmd-3" hidden>cd /home/ubuntu</code>
  </pre>
</div>

  <div>
  <pre><button ng-click="copyToClipboard('cmd-4')" type="button" class="copyButton"><i class="fa fa-files-o"></i></button>
  <code class="language-bash">$ source venv/bin/activate</code>
  <code id="cmd-4" hidden>source venv/bin/activate</code>
  </pre>
</div>

<<<<<<< HEAD
  <p align = "center"><img src="/MizzouCloudDevOps/img/lab9/step6/service_start.png" width=700></p>

=======
>>>>>>> 2da5958087b50fc2d75601a71b7d6a4f2dd757c2
  <h3>5.4.2 Start the Server</h3>
  <ol>
    <li>Update SERVER_PUBLIC_IP</li>
    <li>Run the script below</li>
  </ol>
<div>
  <pre><button ng-click="copyToClipboard('cmd-5')" type="button" class="copyButton"><i class="fa fa-files-o"></i></button>
  <code class="language-bash">$ python3 control_docker.py --key_file secret_key.pem --vm_type server --ip SERVER_PUBLIC_IP --username ubuntu</code>
  <code id="cmd-5" hidden>python3 control_docker.py --key_file secret_key.pem --vm_type server --ip SERVER_PUBLIC_IP --username ubuntu</code>
  </pre>
</div>

<p align="center"><img src="/MizzouCloudDevOps/img/lab9/step6/Running_control_docker_for_server.png" width=700></p>
<p align="center"><img src="/MizzouCloudDevOps/img/lab9/step6/Pulling_docker_img.png" width=700></p>
<p align="center"><img src="/MizzouCloudDevOps/img/lab9/step6/Running_app_as_a_server.png" width=700></p>

  <blockquote style = "background-color: #e5f5e0 ; margin: 0 ; font-family: calibri;border: 2px solid #00441b; border-left: 6px solid #00441b; padding : 0.5em; border-radius: 8px; background-style:discrete" >
    <font size = 3 color = green> <b> Note: </b></font><font color = 'black'>
    If the step fails, recheck the parameters and retry again.
    </font> </blockquote> 
  <h3>5.4.3 Run Client Workflows</h3>
  <ol>
    <li>Navigate to your Client VM tab in your browser</li>
    <li>Run the following commands to activate the virtual environment.
    <div>
  <pre><button ng-click="copyToClipboard('cmd-3')" type="button" class="copyButton"><i class="fa fa-files-o"></i></button>
  <code class="language-bash">$ cd /home/ubuntu</code>
  <code id="cmd-3" hidden>cd /home/ubuntu</code>
  </pre>
</div>
    <div>
  <pre><button ng-click="copyToClipboard('cmd-4')" type="button" class="copyButton"><i class="fa fa-files-o"></i></button>
  <code class="language-bash">$ source venv/bin/activate</code>
  <code id="cmd-4" hidden>source venv/bin/activate</code>
  </pre>
</div>
    
    
    <li>Choose a workflow from four bioinformatic workflows: FastQC, Read Alignment, InDel Calling/SNP Calling</code></li>
    <li>Update SERVER_PRIVATE_IP and CLIENT_PUBLIC_IP</li>
    <li>Run the script</li>
  </ol><br>
  
  <h4 style="color: black;">FastQC Example</h4>

<div>
  <pre><button ng-click="copyToClipboard('cmd-6')" type="button" class="copyButton"><i class="fa fa-files-o"></i></button>
  <code class="language-bash">$ python3 control_docker.py --key_file secret_key.pem --vm_type client --ip CLIENT_PUBLIC_IP --username ubuntu --policy_host SERVER_PRIVATE_IP --server_app_host SERVER_PRIVATE_IP --workflow sequence_quality --dataset_file /root/SRR2584863_0.1_1.fastq --parameters quiet --threads 2</code>
  <code id="cmd-6" hidden>python3 control_docker.py --key_file secret_key.pem --vm_type client --ip CLIENT_PUBLIC_IP --username ubuntu --policy_host SERVER_PRIVATE_IP --server_app_host SERVER_PRIVATE_IP --workflow sequence_quality --dataset_file /root/SRR2584863_0.1_1.fastq --parameters quiet --threads 2</code>
  </pre>
</div>

<p align="center"><img src="/MizzouCloudDevOps/img/lab9/step6/fastqc.png" width=700></p><br>
  <h4 style="color: black;">Read Alignment Example</h4>
<div>
  <pre><button ng-click="copyToClipboard('cmd-7')" type="button" class="copyButton"><i class="fa fa-files-o"></i></button>
  <code class="language-bash">$ python3 control_docker.py --key_file secret_key.pem --vm_type client --ip CLIENT_PUBLIC_IP --username ubuntu --policy_host SERVER_PRIVATE_IP --server_app_host SERVER_PRIVATE_IP --workflow read_alignment --reference_genome /root/ecoli_reference.fa --paired_file_1 /root/SRR2584863_0.1_1.fastq --paired_file_2 /root/SRR2584863_0.1_2.fastq</code>
  <code id="cmd-7" hidden>python3 control_docker.py --key_file secret_key.pem --vm_type client --ip CLIENT_PUBLIC_IP --username ubuntu --policy_host SERVER_PRIVATE_IP --server_app_host SERVER_PRIVATE_IP --workflow read_alignment --reference_genome /root/ecoli_reference.fa --paired_file_1 /root/SRR2584863_0.1_1.fastq --paired_file_2 /root/SRR2584863_0.1_2.fastq</code>
  </pre>
</div>
<p align="center"><img src="/MizzouCloudDevOps/img/lab9/step6/read_alignment.png" width=700></p><br>
  <h4 style="color: black;">Variant Calling Examples (GATK)</h4>

  <h5>Indel Calling</h5>
<div>
  <pre><button ng-click="copyToClipboard('cmd-8')" type="button" class="copyButton"><i class="fa fa-files-o"></i></button>
  <code class="language-bash">$ python3 control_docker.py --key_file secret_key.pem --vm_type client --ip CLIENT_PUBLIC_IP --username ubuntu --policy_host SERVER_PRIVATE_IP --server_app_host SERVER_PRIVATE_IP --workflow indel_calling --reference_genome /root/ecoli_reference.fa --dataset_file /root/example_0.1.rg.bam</code>
  <code id="cmd-8" hidden>python3 control_docker.py --key_file secret_key.pem --vm_type client --ip CLIENT_PUBLIC_IP --username ubuntu --policy_host SERVER_PRIVATE_IP --server_app_host SERVER_PRIVATE_IP --workflow indel_calling --reference_genome /root/ecoli_reference.fa --dataset_file /root/example_0.1.rg.bam</code>
  </pre>
</div>
<p align="center"><img src="/MizzouCloudDevOps/img/lab9/step6/indel_calling.png" width=700></p>
  <h5>SNP Calling</h5>
<div>
  <pre><button ng-click="copyToClipboard('cmd-9')" type="button" class="copyButton"><i class="fa fa-files-o"></i></button>
  <code class="language-bash">$ python3 control_docker.py --key_file secret_key.pem --vm_type client --ip CLIENT_PUBLIC_IP --username ubuntu --policy_host SERVER_PRIVATE_IP --server_app_host SERVER_PRIVATE_IP --workflow snp_calling --reference_genome /root/ecoli_reference.fa --dataset_file /root/example_0.1.rg.bam</code>
  <code id="cmd-9" hidden>python3 control_docker.py --key_file secret_key.pem --vm_type client --ip CLIENT_PUBLIC_IP --username ubuntu --policy_host SERVER_PRIVATE_IP --server_app_host SERVER_PRIVATE_IP --workflow snp_calling --reference_genome /root/ecoli_reference.fa --dataset_file /root/example_0.1.rg.bam</code>
  </pre>
</div>

<p align="center"><img src="/MizzouCloudDevOps/img/lab9/step6/snp_calling.png" width=700></p>
<blockquote style = "background-color: #e5f5e0 ; margin: 0 ; font-family: calibri;border: 2px solid #00441b; border-left: 6px solid #00441b; padding : 0.5em; border-radius: 8px; background-style:discrete" >
    <font size = 3 color = green> <b> Note: </b></font><font color = 'black'>
    If the step fails, recheck the parameters and retry again.
    </font> </blockquote> 

  <!-- <h2>5.5 Execution Steps</h2>

  <h3>5.5.1 Start the Server</h3>
  <ol>
    <li>Update SERVER_IP</li>
    <li>Run the script (see above)</li>
  </ol>

  <h3>5.5.2 Run a Client Workflow</h3>
  <ol>
    <li>Choose a workflow in <code>control_docker.py</code></li>
    <li>Update IP and key path</li>
    <li>Run the script</li>
  </ol>

  <h3>Monitor Progress</h3>
  <ul>
    <li>Real-time updates in terminal</li>
    <li>Server logs: attestation and policy enforcement</li>
    <li>Client logs: workflow execution</li>
  </ul>

  <h3>Customizing Workflows</h3>
  <ul>
    <li><strong>FastQC:</strong> Adjust threads and parameters</li>
    <li><strong>Read Alignment:</strong> Change input file paths</li>
    <li><strong>Variant Calling:</strong> Tweak filtering thresholds or logic</li>
  </ul> -->

  <!--<h2>Troubleshooting</h2>

  <h3>Connection Issues</h3>
  <pre><code>chmod 600 PATH_TO_SSH_KEY</code></pre>
  <p>Ensure firewall/security group rules allow SSH.</p>
  -->

  <h2>5.6 Conclusion</h2>
  <p>This pipeline showcases the transformative potential of <strong>Confidential Computing</strong> in safeguarding sensitive genomic data during critical stages of analysis. By leveraging trusted execution environments (TEEs), the system ensures <strong>data integrity, confidentiality, and policy enforcement</strong> throughout the workflow. The integrated <strong>Python controller</strong> offers a powerful, extensible interface for orchestrating complex bioinformatics pipelines—bridging security, scalability, and automation across distributed cloud or edge ecosystems.</p>



<script type="text/javascript">
  var _self="undefined"!=typeof window?window:"undefined"!=typeof WorkerGlobalScope&&self instanceof WorkerGlobalScope?self:{},Prism=function(){var e=/\blang(?:uage)?-(\w+)\b/i,t=0,n=_self.Prism={util:{encode:function(e){return e instanceof a?new a(e.type,n.util.encode(e.content),e.alias):"Array"===n.util.type(e)?e.map(n.util.encode):e.replace(/&/g,"&amp;").replace(/</g,"&lt;").replace(/\u00a0/g," ")},type:function(e){return Object.prototype.toString.call(e).match(/\[object (\w+)\]/)[1]},objId:function(e){return e.__id||Object.defineProperty(e,"__id",{value:++t}),e.__id},clone:function(e){var t=n.util.type(e);switch(t){case"Object":var a={};for(var r in e)e.hasOwnProperty(r)&&(a[r]=n.util.clone(e[r]));return a;case"Array":return e.map&&e.map(function(e){return n.util.clone(e)})}return e}},languages:{extend:function(e,t){var a=n.util.clone(n.languages[e]);for(var r in t)a[r]=t[r];return a},insertBefore:function(e,t,a,r){r=r||n.languages;var l=r[e];if(2==arguments.length){a=arguments[1];for(var i in a)a.hasOwnProperty(i)&&(l[i]=a[i]);return l}var o={};for(var s in l)if(l.hasOwnProperty(s)){if(s==t)for(var i in a)a.hasOwnProperty(i)&&(o[i]=a[i]);o[s]=l[s]}return n.languages.DFS(n.languages,function(t,n){n===r[e]&&t!=e&&(this[t]=o)}),r[e]=o},DFS:function(e,t,a,r){r=r||{};for(var l in e)e.hasOwnProperty(l)&&(t.call(e,l,e[l],a||l),"Object"!==n.util.type(e[l])||r[n.util.objId(e[l])]?"Array"!==n.util.type(e[l])||r[n.util.objId(e[l])]||(r[n.util.objId(e[l])]=!0,n.languages.DFS(e[l],t,l,r)):(r[n.util.objId(e[l])]=!0,n.languages.DFS(e[l],t,null,r)))}},plugins:{},highlightAll:function(e,t){var a={callback:t,selector:'code[class*="language-"], [class*="language-"] code, code[class*="lang-"], [class*="lang-"] code'};n.hooks.run("before-highlightall",a);for(var r,l=a.elements||document.querySelectorAll(a.selector),i=0;r=l[i++];)n.highlightElement(r,e===!0,a.callback)},highlightElement:function(t,a,r){for(var l,i,o=t;o&&!e.test(o.className);)o=o.parentNode;o&&(l=(o.className.match(e)||[,""])[1],i=n.languages[l]),t.className=t.className.replace(e,"").replace(/\s+/g," ")+" language-"+l,o=t.parentNode,/pre/i.test(o.nodeName)&&(o.className=o.className.replace(e,"").replace(/\s+/g," ")+" language-"+l);var s=t.textContent,u={element:t,language:l,grammar:i,code:s};if(!s||!i)return n.hooks.run("complete",u),void 0;if(n.hooks.run("before-highlight",u),a&&_self.Worker){var c=new Worker(n.filename);c.onmessage=function(e){u.highlightedCode=e.data,n.hooks.run("before-insert",u),u.element.innerHTML=u.highlightedCode,r&&r.call(u.element),n.hooks.run("after-highlight",u),n.hooks.run("complete",u)},c.postMessage(JSON.stringify({language:u.language,code:u.code,immediateClose:!0}))}else u.highlightedCode=n.highlight(u.code,u.grammar,u.language),n.hooks.run("before-insert",u),u.element.innerHTML=u.highlightedCode,r&&r.call(t),n.hooks.run("after-highlight",u),n.hooks.run("complete",u)},highlight:function(e,t,r){var l=n.tokenize(e,t);return a.stringify(n.util.encode(l),r)},tokenize:function(e,t){var a=n.Token,r=[e],l=t.rest;if(l){for(var i in l)t[i]=l[i];delete t.rest}e:for(var i in t)if(t.hasOwnProperty(i)&&t[i]){var o=t[i];o="Array"===n.util.type(o)?o:[o];for(var s=0;s<o.length;++s){var u=o[s],c=u.inside,g=!!u.lookbehind,h=!!u.greedy,f=0,d=u.alias;u=u.pattern||u;for(var p=0;p<r.length;p++){var m=r[p];if(r.length>e.length)break e;if(!(m instanceof a)){u.lastIndex=0;var y=u.exec(m),v=1;if(!y&&h&&p!=r.length-1){var b=r[p+1].matchedStr||r[p+1],k=m+b;if(p<r.length-2&&(k+=r[p+2].matchedStr||r[p+2]),u.lastIndex=0,y=u.exec(k),!y)continue;var w=y.index+(g?y[1].length:0);if(w>=m.length)continue;var _=y.index+y[0].length,P=m.length+b.length;if(v=3,P>=_){if(r[p+1].greedy)continue;v=2,k=k.slice(0,P)}m=k}if(y){g&&(f=y[1].length);var w=y.index+f,y=y[0].slice(f),_=w+y.length,S=m.slice(0,w),O=m.slice(_),j=[p,v];S&&j.push(S);var A=new a(i,c?n.tokenize(y,c):y,d,y,h);j.push(A),O&&j.push(O),Array.prototype.splice.apply(r,j)}}}}}return r},hooks:{all:{},add:function(e,t){var a=n.hooks.all;a[e]=a[e]||[],a[e].push(t)},run:function(e,t){var a=n.hooks.all[e];if(a&&a.length)for(var r,l=0;r=a[l++];)r(t)}}},a=n.Token=function(e,t,n,a,r){this.type=e,this.content=t,this.alias=n,this.matchedStr=a||null,this.greedy=!!r};if(a.stringify=function(e,t,r){if("string"==typeof e)return e;if("Array"===n.util.type(e))return e.map(function(n){return a.stringify(n,t,e)}).join("");var l={type:e.type,content:a.stringify(e.content,t,r),tag:"span",classes:["token",e.type],attributes:{},language:t,parent:r};if("comment"==l.type&&(l.attributes.spellcheck="true"),e.alias){var i="Array"===n.util.type(e.alias)?e.alias:[e.alias];Array.prototype.push.apply(l.classes,i)}n.hooks.run("wrap",l);var o="";for(var s in l.attributes)o+=(o?" ":"")+s+'="'+(l.attributes[s]||"")+'"';return"<"+l.tag+' class="'+l.classes.join(" ")+'" '+o+">"+l.content+"</"+l.tag+">"},!_self.document)return _self.addEventListener?(_self.addEventListener("message",function(e){var t=JSON.parse(e.data),a=t.language,r=t.code,l=t.immediateClose;_self.postMessage(n.highlight(r,n.languages[a],a)),l&&_self.close()},!1),_self.Prism):_self.Prism;var r=document.currentScript||[].slice.call(document.getElementsByTagName("script")).pop();return r&&(n.filename=r.src,document.addEventListener&&!r.hasAttribute("data-manual")&&document.addEventListener("DOMContentLoaded",n.highlightAll)),_self.Prism}();"undefined"!=typeof module&&module.exports&&(module.exports=Prism),"undefined"!=typeof global&&(global.Prism=Prism);
</script>
  
<script type="text/javascript">
  !function(e){var t={variable:[{pattern:/\$?\(\([\w\W]+?\)\)/,inside:{variable:[{pattern:/(^\$\(\([\w\W]+)\)\)/,lookbehind:!0},/^\$\(\(/],number:/\b-?(?:0x[\dA-Fa-f]+|\d*\.?\d+(?:[Ee]-?\d+)?)\b/,operator:/--?|-=|\+\+?|\+=|!=?|~|\*\*?|\*=|\/=?|%=?|<<=?|>>=?|<=?|>=?|==?|&&?|&=|\^=?|\|\|?|\|=|\?|:/,punctuation:/\(\(?|\)\)?|,|;/}},{pattern:/\$\([^)]+\)|`[^`]+`/,inside:{variable:/^\$\(|^`|\)$|`$/}},/\$(?:[a-z0-9_#\?\*!@]+|\{[^}]+\})/i]};e.languages.bash={shebang:{pattern:/^#!\s*\/bin\/bash|^#!\s*\/bin\/sh/,alias:"important"},comment:{pattern:/(^|[^"{\\])#.*/,lookbehind:!0},string:[{pattern:/((?:^|[^<])<<\s*)(?:"|')?(\w+?)(?:"|')?\s*\r?\n(?:[\s\S])*?\r?\n\2/g,lookbehind:!0,inside:t},{pattern:/(["'])(?:\\\\|\\?[^\\])*?\1/g,inside:t}],variable:t.variable,"function":{pattern:/(^|\s|;|\||&)(?:alias|apropos|apt-get|aptitude|aspell|awk|basename|bash|bc|bg|builtin|bzip2|cal|cat|cd|cfdisk|chgrp|chmod|chown|chroot|chkconfig|cksum|clear|cmp|comm|command|cp|cron|crontab|csplit|cut|date|dc|dd|ddrescue|df|diff|diff3|dig|dir|dircolors|dirname|dirs|dmesg|du|egrep|eject|enable|env|ethtool|eval|exec|expand|expect|export|expr|fdformat|fdisk|fg|fgrep|file|find|fmt|fold|format|free|fsck|ftp|fuser|gawk|getopts|git|grep|groupadd|groupdel|groupmod|groups|gzip|hash|head|help|hg|history|hostname|htop|iconv|id|ifconfig|ifdown|ifup|import|install|jobs|join|kill|killall|less|link|ln|locate|logname|logout|look|lpc|lpr|lprint|lprintd|lprintq|lprm|ls|lsof|make|man|mkdir|mkfifo|mkisofs|mknod|more|most|mount|mtools|mtr|mv|mmv|nano|netstat|nice|nl|nohup|notify-send|nslookup|open|op|passwd|paste|pathchk|ping|pkill|popd|pr|printcap|printenv|printf|ps|pushd|pv|pwd|quota|quotacheck|quotactl|ram|rar|rcp|read|readarray|readonly|reboot|rename|renice|remsync|rev|rm|rmdir|rsync|screen|scp|sdiff|sed|seq|service|sftp|shift|shopt|shutdown|sleep|slocate|sort|source|split|ssh|stat|strace|su|sudo|sum|suspend|sync|tail|tar|tee|test|time|timeout|times|touch|top|traceroute|trap|tr|tsort|tty|type|ulimit|umask|umount|unalias|uname|unexpand|uniq|units|unrar|unshar|uptime|useradd|userdel|usermod|users|uuencode|uudecode|v|vdir|vi|vmstat|wait|watch|wc|wget|whereis|which|who|whoami|write|xargs|xdg-open|yes|zip)(?=$|\s|;|\||&)/,lookbehind:!0},keyword:{pattern:/(^|\s|;|\||&)(?:let|:|\.|if|then|else|elif|fi|for|break|continue|while|in|case|function|select|do|done|until|echo|exit|return|set|declare)(?=$|\s|;|\||&)/,lookbehind:!0},"boolean":{pattern:/(^|\s|;|\||&)(?:true|false)(?=$|\s|;|\||&)/,lookbehind:!0},operator:/&&?|\|\|?|==?|!=?|<<<?|>>|<=?|>=?|=~/,punctuation:/\$?\(\(?|\)\)?|\.\.|[{}[\];]/};var a=t.variable[1].inside;a["function"]=e.languages.bash["function"],a.keyword=e.languages.bash.keyword,a.boolean=e.languages.bash.boolean,a.operator=e.languages.bash.operator,a.punctuation=e.languages.bash.punctuation}(Prism);
</script>


</body>

</html>
